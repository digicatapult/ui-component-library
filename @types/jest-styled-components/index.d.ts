declare module 'jest-styled-components/serializer'
