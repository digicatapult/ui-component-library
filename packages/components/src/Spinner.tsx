import React from 'react'
import styled from 'styled-components'
import sx, {SxProp} from './sx'
import {ComponentProps} from './utils/types'

const sizeMap = {
  small: '16px',
  medium: '32px',
  large: '64px',
}

export interface SpinnerInternalProps {
  /** Sets the width and height of the spinner. */
  size?: keyof typeof sizeMap
}

function Spinner({size: sizeKey = 'medium', ...props}: SpinnerInternalProps) {
  const size = sizeMap[sizeKey]

  return (
    <svg height={size} width={size} viewBox="0 0 16 16" fill="none" {...props}>
      <circle
        cx="8"
        cy="8"
        r="7"
        stroke="currentColor"
        strokeOpacity="0.25"
        strokeWidth="2"
        vectorEffect="non-scaling-stroke"
      />
      <path
        d="M15 8a7.002 7.002 0 00-7-7"
        stroke="currentColor"
        strokeWidth="2"
        strokeLinecap="round"
        vectorEffect="non-scaling-stroke"
      />
    </svg>
  )
}

const StyledSpinner = styled(Spinner)<SxProp>`
  @keyframes rotate-keyframes {
    100% {
      transform: rotate(360deg);
    }
  }

  animation: rotate-keyframes 1s linear infinite;

  ${sx}
`

StyledSpinner.displayName = 'Spinner'

export type SpinnerProps = ComponentProps<typeof StyledSpinner>
export default StyledSpinner
