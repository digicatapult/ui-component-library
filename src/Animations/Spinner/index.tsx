import React from 'react'
import styled from 'styled-components'

export type SpinnerProps = {
  styles?: React.CSSProperties
  text?: string
  color?: string
  size?: 'small' | 'medium' | 'large' | undefined
}

const sizeToPixel = {
  small: '10px',
  medium: '30px',
  large: '50px',
}

const Spinner: React.FC<SpinnerProps> = ({ text, ...props }) => {
  return (
    <Wrapper>
      <Animation {...props} />
      <Label>{text}</Label>
    </Wrapper>
  )
}

const Animation = styled('div')`
  border: 10px solid #f3f3f3;
  border-radius: 50%;
  border-top: 10px solid ${(props: SpinnerProps) => props.color || '#3498db'};
  width: ${(props: SpinnerProps) => sizeToPixel[props.size || 'small']};
  height: ${(props: SpinnerProps) => sizeToPixel[props.size || 'small']};
  margin: auto;
  padding: auto;
  -webkit-animation: spin 1.75s linear infinite; /* Safari */
  animation: spin 1.75s linear infinite;
  }

  /* Safari */
  @-webkit-keyframes spin {
  0% { -webkit-transform: rotate(0deg); }
  100% { -webkit-transform: rotate(360deg); }
  }

  @keyframes spin {
  0% { transform: rotate(0deg); }
  100% { transform: rotate(360deg); }
  }
`
const Wrapper = styled.div`
  align-self: center;
  justify-self: center;
`

const Label = styled.h2`
  text-align: center;
`

export default Spinner
