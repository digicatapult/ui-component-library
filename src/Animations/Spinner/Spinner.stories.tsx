import React from 'react'

import Spinner, { SpinnerProps } from './index.js'

export default {
  title: 'Components/Animations/Spinner',
  component: Spinner,
  argTypes: {
    text: {
      control: 'text',
    },
    color: {
      control: {
        type: 'select',
      },
      options: ['#a7a7a7', '#fff', '#00ff00', '#33e58c', 'red'],
    },
  },
}

const DefaultStoryTemplate = (args: SpinnerProps) => {
  return <Spinner {...args} />
}

export const Default = DefaultStoryTemplate.bind({})
Default.args = {
  color: '#a7a7a7',
}

export const Small = DefaultStoryTemplate.bind({})
Small.args = {
  size: 'small',
  text: 'this is small spinner example',
}

export const Medium = DefaultStoryTemplate.bind({})
Medium.args = {
  size: 'medium',
  text: 'this is medium spinner example',
}
export const Large = DefaultStoryTemplate.bind({})
Large.args = {
  size: 'large',
  text: 'this is large spinner example',
}

export const HyProof = DefaultStoryTemplate.bind({})
HyProof.args = {
  color: '#33e58c',
  text: 'this is hyproof spinner example',
}
