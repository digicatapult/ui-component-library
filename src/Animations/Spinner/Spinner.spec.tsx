import React from 'react'
import renderer from 'react-test-renderer'
import 'jest-styled-components'

import Spinner from './index.js'

describe('Spinner', () => {
  test('default', () => {
    const instance = renderer.create(<Spinner />)
    const tree = instance.toJSON()

    expect(tree).toMatchSnapshot()
  })

  test('with color and text props', () => {
    const instance = renderer.create(
      <Spinner color="#ff00ff" text="this is a test" />,
    )
    const tree = instance.toJSON()

    expect(tree).toMatchSnapshot()
  })

  test('renders a small spinner', () => {
    const instance = renderer.create(<Spinner size="small" />)

    const tree = instance.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('renders a medium spinner', () => {
    const instance = renderer.create(<Spinner size="medium" />)

    const tree = instance.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('renders a large spinner', () => {
    const instance = renderer.create(<Spinner size="large" />)

    const tree = instance.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
