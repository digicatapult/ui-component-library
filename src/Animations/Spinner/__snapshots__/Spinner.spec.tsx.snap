// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spinner default 1`] = `
.c1 {
  border: 5px solid #f3f3f3;
  border-radius: 50%;
  border-top: 5px solid #3498db;
  width: 10px;
  height: 10px;
  margin: auto;
  padding: auto;
  animation: spin 1.75s linear infinite;
}

.c0 {
  align-self: center;
  justify-self: center;
}

.c2 {
  text-align: center;
}

<div
  className="c0"
>
  <div
    className="c1"
  />
  <div
    className="c2"
  />
</div>
`;

exports[`Spinner renders a large spinner 1`] = `
.c1 {
  border: 5px solid #f3f3f3;
  border-radius: 50%;
  border-top: 5px solid #3498db;
  width: 50px;
  height: 50px;
  margin: auto;
  padding: auto;
  animation: spin 1.75s linear infinite;
}

.c0 {
  align-self: center;
  justify-self: center;
}

.c2 {
  text-align: center;
}

<div
  className="c0"
>
  <div
    className="c1"
    size="large"
  />
  <div
    className="c2"
  />
</div>
`;

exports[`Spinner renders a medium spinner 1`] = `
.c1 {
  border: 5px solid #f3f3f3;
  border-radius: 50%;
  border-top: 5px solid #3498db;
  width: 30px;
  height: 30px;
  margin: auto;
  padding: auto;
  animation: spin 1.75s linear infinite;
}

.c0 {
  align-self: center;
  justify-self: center;
}

.c2 {
  text-align: center;
}

<div
  className="c0"
>
  <div
    className="c1"
    size="medium"
  />
  <div
    className="c2"
  />
</div>
`;

exports[`Spinner renders a small spinner 1`] = `
.c1 {
  border: 5px solid #f3f3f3;
  border-radius: 50%;
  border-top: 5px solid #3498db;
  width: 10px;
  height: 10px;
  margin: auto;
  padding: auto;
  animation: spin 1.75s linear infinite;
}

.c0 {
  align-self: center;
  justify-self: center;
}

.c2 {
  text-align: center;
}

<div
  className="c0"
>
  <div
    className="c1"
    size="small"
  />
  <div
    className="c2"
  />
</div>
`;

exports[`Spinner with color and text props 1`] = `
.c1 {
  border: 5px solid #f3f3f3;
  border-radius: 50%;
  border-top: 5px solid #ff00ff;
  width: 10px;
  height: 10px;
  margin: auto;
  padding: auto;
  animation: spin 1.75s linear infinite;
}

.c0 {
  align-self: center;
  justify-self: center;
}

.c2 {
  text-align: center;
}

<div
  className="c0"
>
  <div
    className="c1"
    color="#ff00ff"
  />
  <div
    className="c2"
  >
    this is a test
  </div>
</div>
`;
